store/decoupled/store.go

```diff
@@ -0,0 +1,161 @@
+package decoupled
+
+import (
+	"crypto/sha256"
+	"io"
+	"sync"
+
+	"github.com/cosmos/cosmos-sdk/store/cachekv"
+	"github.com/cosmos/cosmos-sdk/store/iavl"
+	"github.com/cosmos/cosmos-sdk/store/tracekv"
+	"github.com/cosmos/cosmos-sdk/store/types"
+	// "github.com/cosmos/cosmos-sdk/telemetry"
+
+	// abci "github.com/tendermint/tendermint/abci/types"
+	dbm "github.com/tendermint/tm-db"
+)
+
+var (
+	_ types.KVStore       = (*Store)(nil)
+	_ types.CommitStore   = (*Store)(nil)
+	_ types.CommitKVStore = (*Store)(nil)
+	// _ types.Queryable               = (*Store)(nil)
+	// _ types.StoreWithInitialVersion = (*Store)(nil)
+)
+
+var (
+	versionsPrefix = []byte{0}
+	dataPrefix     = []byte{1}
+	indexPrefix    = []byte{2}
+)
+
+// A store which uses separate data structures for state storage (SS) and state commitments (SC)
+type Store struct {
+	// State commitments layer
+	sc types.CommitKVStore
+	// Direct KV mapping (SS)
+	data dbm.DB
+	// Inverted index of SC values to SS keys
+	inv dbm.DB
+
+	mtx sync.RWMutex
+}
+
+// TODO:
+// version tracking, construction?
+// separate backing DB?
+func NewStore(db dbm.DB, id types.CommitID, lazyLoading bool) (*Store, error) {
+	sc, err := iavl.LoadStore(db, id, lazyLoading)
+	if err != nil {
+		return nil, err
+	}
+	return &Store{
+		sc:   sc,
+		data: dbm.NewPrefixDB(db, dataPrefix),
+		inv:  dbm.NewPrefixDB(db, indexPrefix),
+	}, nil
+}
+
+// implement KVStore
+func (s *Store) Get(key []byte) []byte {
+	val, err := s.data.Get(key)
+	if err != nil {
+		panic(err)
+	}
+	return val
+}
+
+func (s *Store) Has(key []byte) bool {
+	has, err := s.data.Has(key)
+	return err == nil && has
+}
+
+func (s *Store) Set(key []byte, value []byte) {
+	kvHash := sha256.Sum256(append(key, value...))
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	err := s.data.Set(key, value)
+	if err != nil {
+		panic(err.Error())
+	}
+	err = s.inv.Set(kvHash[:], key)
+	if err != nil {
+		panic(err.Error())
+	}
+	s.sc.Set(key, kvHash[:]) // TODO: key or hash(key)?
```

Comment:
Need an `err :=` here. But actually I don't think we need this `Set`, the `key, value` and `hash(key, value), key` will be sufficient

Reply:
This is where the state commitment mapping is set; [`Set()`](https://github.com/vulcanize/cosmos-sdk/blob/82dc0081fdc06fa6cf8bd6a0d87987d27ac7b1ca/store/iavl/store.go#L169) has no return value

Reply:
Got it, we don't need the `err` check below then.